libsrc/arithmetic/man3/im_invert.3
.TH IM_INVERT 3 "11 April 1990"
.SH NAME
im_invert \- inverts an image pointed by an image descriptor.
.SH SYNOPSIS
.B #include <vips/vips.h>

.B int im_invert(in, out)
.br
.B IMAGE *in, *out;
.SH DESCRIPTION
.B im_invert
inverts the image held by image descriptor in and writes the result on the 
image descriptor out.  The function works on FMTUCHAR images only.  
The result at point x of the input image is 255-x at the output image.
Input can have any no of channels.
.br
  This is not a generally useful program -- it is included as an example of a
very simple new-style IO function. Feel the power of the source, Luke!
.SH RETURN VALUE
The function returns 0 on success and -1 on error.
.SH SEE\ ALSO
im_add(3), im_lintra(3), im_multiply(3).
.SH AUTHOR
J. Cupitt \- 21/7/93


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/arithmetic/im_invert.c
/* @(#) Invert a UCHAR image. Very simple new-style VIPS routine. See
 * @(#) im_exptra() for the next level of complexity. This function is not
 * @(#) as quick as it could be - it is intended to be an example rather than
 * @(#) to be useful. This should really be written with im_wrapone().
 * @(#)
 * @(#) int
 * @(#) im_invert( IMAGE *in, IMAGE *out )
 * @(#)
 * @(#) All functions return 0 on success and -1 on error
 * @(#)
 *
 * Copyright: 1990, N. Dessipris.
 *
 * Author: Nicos Dessipris
 * Written on: 12/02/1990
 * Modified on :
 * 7/7/93 JC
 *      - memory leaks fixed
 *      - adapted for partial v2
 *      - ANSIfied
 * 22/2/95 JC
 *	- tidied up again
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */
 
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Invert a REGION. We are given the REGION we should write to, the REGION we
 * should use for input, and the IMAGE we are processing. On entry to
 * invert_gen(), or points to the memory we should write to and ir is blank.
 */
static int
invert_gen( REGION *or, REGION *ir, IMAGE *in )
{
	/* Left, right, top and bottom for the output region.
	 */
	int le = or->valid.left;
	int to = or->valid.top;
	int bo = IM_RECT_BOTTOM( &or->valid );

	int x, y;

	/* Ask for the section of the input image we need to produce this
	 * section of the output image.
	 */
	if( im_prepare( ir, &or->valid ) )
		return( -1 );
	
	/* Loop over output, writing input.
	 */
	for( y = to; y < bo; y++ ) {
		/* Point p and q at the start of the line of pels we must
		 * process this loop.
		 */
		PEL *p = (PEL *) IM_REGION_ADDR( ir, le, y );
		PEL *q = (PEL *) IM_REGION_ADDR( or, le, y );

		/* Loop along the line, processing pels. 
		 * IM_REGION_N_ELEMENTS(region) gives
		 * the number of band elements across a region. By looping to
		 * IM_REGION_N_ELEMENTS() rather than ir->valid.width, we work 
		 * for any number of bands.
		 */
		for( x = 0; x < IM_REGION_N_ELEMENTS( or ); x++ )
			q[x] = 255 - p[x];
	}

        return( 0 );
}

/* Invert IMAGE in to IMAGE out. Any number of bands, unsigned char pels
 * only. See im_exptra() for an example of a VIPS function which can process
 * any input image type.
 */
int
im_invert( IMAGE *in, IMAGE *out )
{
	/* Check args.
	 */
	if( in->Coding != IM_CODING_NONE ) {
		im_error( "im_invert", _( "not uncoded" ) );
		return( -1 );
	}
	if( in->BandFmt != IM_BANDFMT_UCHAR ) {
		im_error( "im_invert", _( "not UCHAR" ) );
		return( -1 );
	}
        if( im_piocheck( in, out ) )
		return( -1 );

	/* Prepare the output header.
	 */
        if( im_cp_desc( out, in ) ) 
		return( -1 );

	/* Set demand hints. Like most one-to-one operations, we work best
	 * with long, thin strips.
	 */
	if( im_demand_hint( out, IM_THINSTRIP, in, NULL ) )
		 return( -1 );

	/* Generate into out. im_start_one() and im_stop_one() are simple
	 * convenience functions provided by VIPS which do the necessary
	 * region creation and destruction for one-image-in
	 * style functions. See im_add(), im_start_many() and im_stop_many() 
	 * for convenience functions for multiple inputs.
	 */
	if( im_generate( out,
		im_start_one, invert_gen, im_stop_one, in, NULL ) )
		return( -1 );
	
	return( 0 );
}
